powertools-idempotency/src/main/java/software/amazon/lambda/powertools/idempotency/internal/IdempotencyHandler.java

```diff
@@ -0,0 +1,159 @@
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.lambda.powertools.idempotency.Idempotency;
+import software.amazon.lambda.powertools.idempotency.exceptions.*;
+import software.amazon.lambda.powertools.idempotency.persistence.BasePersistenceStore;
+import software.amazon.lambda.powertools.idempotency.persistence.DataRecord;
+import software.amazon.lambda.powertools.utilities.JsonConfig;
+
+import java.time.Instant;
+
+import static software.amazon.lambda.powertools.idempotency.persistence.DataRecord.Status.EXPIRED;
+import static software.amazon.lambda.powertools.idempotency.persistence.DataRecord.Status.INPROGRESS;
+
+/**
+ * Internal class that will handle the Idempotency, and use the {@link software.amazon.lambda.powertools.idempotency.persistence.PersistenceStore}
+ * to store the result of previous calls.
+ */
+public class IdempotencyHandler {
+    private static final Logger LOG = LoggerFactory.getLogger(IdempotencyHandler.class);
+    private static final int MAX_RETRIES = 2;
+
+    private final ProceedingJoinPoint pjp;
+    private final JsonNode data;
+    private final BasePersistenceStore persistenceStore;
+
+    public IdempotencyHandler(ProceedingJoinPoint pjp, String functionName, JsonNode payload) {
+        this.pjp = pjp;
+        this.data = payload;
+        persistenceStore = Idempotency.getInstance().getPersistenceStore();
+        persistenceStore.configure(Idempotency.getInstance().getConfig(), functionName);
+    }
+
+    /**
+     * Main entry point for handling idempotent execution of a function.
+     *
+     * @return function response
+     */
+    public Object handle() throws Throwable {
+        // IdempotencyInconsistentStateException can happen under rare but expected cases
+        // when persistent state changes in the small time between put & get requests.
+        // In most cases we can retry successfully on this exception.
+        for (int i = 0; true; i++) {
+            try {
+                return processIdempotency();
+            } catch (IdempotencyInconsistentStateException e) {
+                if (i == MAX_RETRIES) {
+                    throw e;
+                }
+            }
+        }
+    }
+
+    /**
+     * Process the function with idempotency
+     *
+     * @return function response
+     */
+    private Object processIdempotency() throws Throwable {
+        try {
+            // We call saveInProgress first as an optimization for the most common case where no idempotent record
+            // already exists. If it succeeds, there's no need to call getRecord.
+            persistenceStore.saveInProgress(data, Instant.now());
+        } catch (IdempotencyItemAlreadyExistsException iaee) {
+            DataRecord record = getIdempotencyRecord();
+            return handleForStatus(record);
+        } catch (IdempotencyKeyException ike) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Reply:
Won't change this. Same behaviour as python:

https://github.com/awslabs/aws-lambda-powertools-python/blob/24ab7266beacfa8b55f1b0838c27b0983484775e/aws_lambda_powertools/utilities/idempotency/base.py#L104